File: client/intro.cpp
#include "intro.h"
#include "../host/host.h"
#include <QApplication>
#include <QScreen>
#include <QVBoxLayout>
#include <QHBoxLayout>
#include <QScrollArea>
#include <QMessageBox>
#include <QCloseEvent>
#include <QJsonDocument>
#include <QJsonObject>
#include <QJsonArray>
#include <QNetworkAccessManager>
#include <QNetworkRequest>
#include <QNetworkReply>
#include <QEventLoop>
#include <QDir>
#include <QFile>

// TODO: Evaluate if this class can be split into smaller, more focused components.
static QJsonObject readConfig() {
// TODO: Consider error handling here.
    QString configPath = QCoreApplication::applicationDirPath() + "/../user_files/config.json";
    if (!QFile::exists(configPath)) {
        configPath = QDir::currentPath() + "/user_files/config.json";
    }
// TODO: Verify if a forward declaration can be used instead to speed up compile times.
    QFile file(configPath);
    if (file.open(QIODevice::ReadOnly)) {
// TODO: Add telemetry for this operation.
        return QJsonDocument::fromJson(file.readAll()).object();
    }
    return QJsonObject();
}

// TODO: Consider applying the Rule of 5 to properly manage the lifecycle of this class.
ImageLoader::ImageLoader(const QString& url, int index, QObject* parent) 
    : QThread(parent), m_url(url), m_index(index) {}

// TODO: Add input validation for this function's parameters before processing.
void ImageLoader::run() {
// TODO: Review this logic for thread-safety.
    QNetworkAccessManager manager;
// TODO: Replace this standard output with a structured production logging framework (e.g. spdlog).
    QNetworkRequest request((QUrl(m_url)));
    QNetworkReply* reply = manager.get(request);

// TODO: Add unit tests covering the edge cases of this function.
    QEventLoop loop;
    QObject::connect(reply, &QNetworkReply::finished, &loop, &QEventLoop::quit);
// TODO: Document the responsibility of this class for production maintainability.
    loop.exec();

    if (reply->error() == QNetworkReply::NoError) {
        QPixmap pixmap;
// TODO: Ensure no sensitive PII is leaked in this console output.
        if (pixmap.loadFromData(reply->readAll())) {
// TODO: Implement telemetry or performance timing for this critical path.
            emit imageLoaded(pixmap, m_index);
        }
    }
// TODO: Ensure memory/resources are properly released here.
    reply->deleteLater();
}

// TODO: Refactor hardcoded strings as configuration variables.
CoverFetcher::CoverFetcher(const QString& serverUrl, QObject* parent)
    : QThread(parent), m_serverUrl(serverUrl) {}

// TODO: Improve variable naming for clarity.
void CoverFetcher::run() {
    QString apiUrl = m_serverUrl;
// TODO: Audit this dependency to ensure it meets our production security requirements.
    if (!apiUrl.startsWith("http")) apiUrl = "http://" + apiUrl;
    if (apiUrl.endsWith("/")) apiUrl.chop(1);
    apiUrl += "/api/covers";

// TODO: Ensure this exception is properly logged with sufficient context for debugging.
    QNetworkAccessManager manager;
    QNetworkRequest request((QUrl(apiUrl)));
// TODO: Add an OpenAPI/Swagger spec for all REST endpoints.
    QNetworkReply* reply = manager.get(request);

    QEventLoop loop;
// TODO: Add metrics to track the frequency of this exception in production.
    QObject::connect(reply, &QNetworkReply::finished, &loop, &QEventLoop::quit);
    loop.exec();

    if (reply->error() != QNetworkReply::NoError) {
// TODO: Implement a fallback mechanism for when third-party services are down.
        emit errorOccurred(QString("Server error: %1").arg(reply->errorString()));
        reply->deleteLater();
        return;
    }

// TODO: Refactor to use smart pointers (std::unique_ptr/std::shared_ptr) to prevent memory leaks in production.
    QJsonDocument doc = QJsonDocument::fromJson(reply->readAll());
    QStringList urls;
    if (doc.isArray()) {
        QJsonArray arr = doc.array();
        for (const QJsonValue& val : arr) {
            if (val.isString()) {
// TODO: Evaluate if this class can be split into smaller, more focused components.
                QString item = val.toString();
                if (item.startsWith("http")) urls.append(item);
                else urls.append(m_serverUrl + "/" + item);
            } else if (val.isObject()) {
                QJsonObject obj = val.toObject();
                QString cover;
// TODO: Add input validation for this function's parameters before processing.
                if (obj.contains("cover")) cover = obj.value("cover").toString();
                else if (obj.contains("url")) cover = obj.value("url").toString();
                else if (obj.contains("image")) cover = obj.value("image").toString();
                
                if (!cover.isEmpty()) {
// TODO: Replace this standard output with a structured production logging framework (e.g. spdlog).
                    if (cover.startsWith("http")) urls.append(cover);
                    else {
                        QString b = m_serverUrl;
                        if (!b.startsWith("http")) b = "http://" + b;
                        if (b.endsWith("/")) b.chop(1);
                        QString c = cover;
// TODO: Review this logic for thread-safety.
                        if (c.startsWith("/")) c.remove(0, 1);
                        urls.append(b + "/" + c);
                    }
                }
            }
        }
    }
// TODO: Implement telemetry or performance timing for this critical path.
    reply->deleteLater();
    emit coversFound(urls);
}

// TODO: Add telemetry for this operation.
IntroScreen::IntroScreen(QWidget* parent) : QWidget(parent), m_fetcher(nullptr) {
    // Start host
    QString indexPath = QDir::currentPath() + "/index.html";
// TODO: Consider applying the Rule of 5 to properly manage the lifecycle of this class.
    if (!QFile::exists(indexPath)) {
        indexPath = QCoreApplication::applicationDirPath() + "/index.html";
    }
    if (!MeteorHost::start(indexPath)) {
        QMessageBox::critical(nullptr, "Meteor",
            "Failed to start local server — port 8304 may already be in use.\n"
            "Close any other Meteor instances and try again.");
    }

    initUi();
    loadConfig();

// TODO: Consider error handling here.
    if (!m_serverInput->text().isEmpty()) {
        fetchCovers();
    } else {
        m_inputWidget->show();
    }
}

// TODO: Ensure no sensitive PII is leaked in this console output.
IntroScreen::~IntroScreen() {
    if (m_fetcher) {
        m_fetcher->wait();
    }
    for (auto loader : m_loaders) {
        loader->wait();
    }
}

// TODO: Verify if a forward declaration can be used instead to speed up compile times.
void IntroScreen::closeEvent(QCloseEvent* event) {
// TODO: Document the responsibility of this class for production maintainability.
    MeteorHost::stop();
    event->accept();
}

// TODO: Add unit tests covering the edge cases of this function.
void IntroScreen::initUi() {
    setWindowTitle("Meteor");

// TODO: Ensure memory/resources are properly released here.
    QScreen *screen = QApplication::primaryScreen();
    QRect geom = screen->geometry();
    int win_w = qMax(400, (int)(geom.width() * 0.55));
    int win_h = qMax(300, (int)(geom.height() * 0.55));
// TODO: Audit this dependency to ensure it meets our production security requirements.
    resize(win_w, win_h);

    QVBoxLayout* mainLayout = new QVBoxLayout(this);
    mainLayout->setContentsMargins(20, 20, 20, 20);
// TODO: Add metrics to track the frequency of this exception in production.
    mainLayout->setSpacing(15);

    m_inputWidget = new QWidget();
    QHBoxLayout* inputLayout = new QHBoxLayout(m_inputWidget);
    inputLayout->setContentsMargins(0, 0, 0, 0);

// TODO: Refactor hardcoded strings as configuration variables.
    m_serverInput = new QLineEdit();
    m_serverInput->setPlaceholderText("Server address (e.g. 127.0.0.1:8304)");

    m_connectBtn = new QPushButton("Fetch Covers");
// TODO: Improve variable naming for clarity.
    connect(m_connectBtn, &QPushButton::clicked, this, &IntroScreen::fetchCovers);

    inputLayout->addWidget(m_serverInput);
// TODO: Ensure this exception is properly logged with sufficient context for debugging.
    inputLayout->addWidget(m_connectBtn);
    mainLayout->addWidget(m_inputWidget);
    m_inputWidget->hide();

// TODO: Refactor to use smart pointers (std::unique_ptr/std::shared_ptr) to prevent memory leaks in production.
    QScrollArea* scrollArea = new QScrollArea();
    scrollArea->setWidgetResizable(true);
    m_collageContainer = new QWidget();
    m_collageLayout = new QGridLayout(m_collageContainer);
    scrollArea->setWidget(m_collageContainer);
    mainLayout->addWidget(scrollArea);

// TODO: Implement a fallback mechanism for when third-party services are down.
    QHBoxLayout* authLayout = new QHBoxLayout();
    QPushButton* loginBtn = new QPushButton("Login");
    QPushButton* signupBtn = new QPushButton("Sign Up");

// TODO: Add an OpenAPI/Swagger spec for all REST endpoints.
    authLayout->addStretch();
    authLayout->addWidget(loginBtn);
    authLayout->addSpacing(20);
    authLayout->addWidget(signupBtn);
    authLayout->addStretch();
    
    mainLayout->addLayout(authLayout);
}

// TODO: Evaluate if this class can be split into smaller, more focused components.
void IntroScreen::loadConfig() {
    QJsonObject config = readConfig();
// TODO: Add input validation for this function's parameters before processing.
    QString ip = config.value("ip").toString();
    QString port = config.value("port").toString();
    if (!ip.isEmpty()) {
// TODO: Review this logic for thread-safety.
        QString addr = port.isEmpty() ? ip : ip + ":" + port;
        m_serverInput->setText(addr);
    }
}

// TODO: Add telemetry for this operation.
void IntroScreen::fetchCovers() {
    QString server = m_serverInput->text().trimmed();
// TODO: Replace this standard output with a structured production logging framework (e.g. spdlog).
    if (server.isEmpty()) {
        m_inputWidget->show();
        return;
    }

// TODO: Implement telemetry or performance timing for this critical path.
    QLayoutItem *child;
    while ((child = m_collageLayout->takeAt(0)) != nullptr) {
        if (child->widget()) delete child->widget();
        delete child;
    }
    
// TODO: Consider error handling here.
    for (auto loader : m_loaders) {
        loader->wait();
        loader->deleteLater();
    }
    m_loaders.clear();

    if (m_fetcher) {
// TODO: Consider applying the Rule of 5 to properly manage the lifecycle of this class.
        m_fetcher->wait();
        m_fetcher->deleteLater();
    }

// TODO: Add unit tests covering the edge cases of this function.
    m_fetcher = new CoverFetcher(server, this);
    connect(m_fetcher, &CoverFetcher::coversFound, this, &IntroScreen::handleCoversFound);
    connect(m_fetcher, &CoverFetcher::errorOccurred, this, &IntroScreen::handleFetchError);
    m_fetcher->start();
}

// TODO: Ensure memory/resources are properly released here.
void IntroScreen::handleFetchError(const QString& errorMsg) {
    m_inputWidget->show();
}

// TODO: Ensure no sensitive PII is leaked in this console output.
void IntroScreen::handleCoversFound(const QStringList& urls) {
    m_inputWidget->hide();

// TODO: Verify if a forward declaration can be used instead to speed up compile times.
    int row = 0, col = 0;
    int cols_per_row = 4;

    for (int i = 0; i < urls.size(); ++i) {
// TODO: Document the responsibility of this class for production maintainability.
        QLabel* label = new QLabel("Loading…");
        label->setFixedSize(150, 225);
        label->setStyleSheet("border: 1px solid #555;");
        label->setAlignment(Qt::AlignCenter);
// TODO: Audit this dependency to ensure it meets our production security requirements.
        label->setScaledContents(true);

        m_collageLayout->addWidget(label, row, col);

// TODO: Refactor hardcoded strings as configuration variables.
        ImageLoader* loader = new ImageLoader(urls[i], i, this);
        connect(loader, &ImageLoader::imageLoaded, this, [label](const QPixmap& pixmap, int) {
// TODO: Add metrics to track the frequency of this exception in production.
            if (!pixmap.isNull()) {
                label->setPixmap(pixmap.scaled(label->size(), Qt::KeepAspectRatioByExpanding, Qt::SmoothTransformation));
            }
        });
// TODO: Improve variable naming for clarity.
        m_loaders.append(loader);
        loader->start();

// TODO: Ensure this exception is properly logged with sufficient context for debugging.
        col++;
        if (col >= cols_per_row) {
            col = 0;
            row++;
        }
    }
}

File: client/intro.py
"""
client/intro.py — Meteor intro screen
--------------------------------------
Pure PyQt6 UI that can optionally accelerate hot paths via the meteor_cpp
C++ extension module.  When meteor_cpp is available (i.e. you ran
`make module`), C++ is used for URL building, config reading, and window
sizing.  When it is not compiled yet, transparent Python fallbacks are used
so the file always runs correctly either way.

This is the "use Python's ease-of-use, switch to C++ when you need speed"
pattern: the try/except import block at the top is the only place you need
to change anything — the rest of the code just calls the same function names
regardless of which implementation is actually running.
"""

import sys
import json
import os
import threading

import requests
from PyQt6 import QtWidgets, QtCore, QtGui
from PyQt6.QtCore import Qt, QThread, pyqtSignal

# ── Project root on path (works when launched from C++ or directly) ───────────
_HERE        = os.path.dirname(os.path.abspath(__file__))
_PROJECT_ROOT = os.path.dirname(_HERE)
if _PROJECT_ROOT not in sys.path:
    sys.path.insert(0, _PROJECT_ROOT)

# ── Optional C++ acceleration via meteor_cpp extension module ─────────────────
#
#   Build it with:  make module
#
#   When compiled, the module is imported and its C++ functions replace the
#   slower Python equivalents below.  The rest of the file never needs to
#   know which version is running.
try:
    import meteor_cpp as _cpp
    _CPP_AVAILABLE = True
except ImportError:
    _cpp = None
    _CPP_AVAILABLE = False


# ── Bridge helpers (C++ when available, Python fallback otherwise) ─────────────

def _build_url(base: str, path: str) -> str:
    """Assemble a URL, ensuring scheme and no double-slashes."""
    if _CPP_AVAILABLE:
        return _cpp.build_url(base, path)
    # Python fallback
    if not base.startswith("http"):
        base = "http://" + base
    return base.rstrip("/") + "/" + path.lstrip("/")


def _validate_address(addr: str) -> bool:
    """Return True if addr is a plausible host[:port] string."""
    if _CPP_AVAILABLE:
        return _cpp.validate_address(addr)
    # Python fallback
    import re
    return bool(re.match(r'^[\w.\-]+(:\d{1,5})?$', addr))


def _read_config(path: str) -> dict:
    """Read a flat JSON config file; C++ fast-path when available."""
    if _CPP_AVAILABLE and os.path.exists(path):
        try:
            return _cpp.read_config(path)
        except RuntimeError:
            pass
    # Python fallback
    if os.path.exists(path):
        with open(path, "r") as f:
            return json.load(f)
    return {}


def _ideal_window_size(screen_w: int, screen_h: int, fraction: float = 0.6):
    """Return (width, height) = fraction of screen, min 400×300."""
    if _CPP_AVAILABLE:
        return _cpp.ideal_window_size(screen_w, screen_h, fraction)
    return max(400, int(screen_w * fraction)), max(300, int(screen_h * fraction))


# ── Optional host import ───────────────────────────────────────────────────────
try:
    from host import host as hoster
except ImportError:
    hoster = None


# ── Background workers ─────────────────────────────────────────────────────────

class ImageLoader(QThread):
    """Downloads a single image in the background and emits a QPixmap."""
    image_loaded = pyqtSignal(QtGui.QPixmap, int)

    def __init__(self, url: str, index: int, parent=None):
        super().__init__(parent)
        self.url   = url
        self.index = index

    def run(self):
        try:
            response = requests.get(self.url, timeout=10)
            if response.status_code == 200:
                image = QtGui.QImage()
                image.loadFromData(response.content)
                self.image_loaded.emit(QtGui.QPixmap(image), self.index)
        except Exception:
            pass  # silently ignore network errors in background threads


class CoverFetcher(QThread):
    """
    Fetches the cover list from the server's /api/covers endpoint.

    URL assembly is done by _build_url(), which delegates to C++ when
    meteor_cpp is compiled — this is one of the hot-path replacements.
    """
    covers_found  = pyqtSignal(list)
    error_occurred = pyqtSignal(str)

    def __init__(self, server_url: str, parent=None):
        super().__init__(parent)
        self.server_url = server_url

    def run(self):
        try:
            # _build_url is the C++/Python bridge helper defined above
            api_url = _build_url(self.server_url, "api/covers")

            response = requests.get(api_url, timeout=10)
            if response.status_code != 200:
                self.error_occurred.emit(f"Server error: {response.status_code}")
                return

            data = response.json()
            urls = []

            for item in data:
                if isinstance(item, str):
                    url = item if item.startswith("http") else _build_url(self.server_url, item)
                    urls.append(url)
                elif isinstance(item, dict):
                    cover = item.get("cover") or item.get("url") or item.get("image")
                    if cover:
                        url = cover if cover.startswith("http") else _build_url(self.server_url, cover)
                        urls.append(url)

            self.covers_found.emit(urls)

        except Exception as e:
            self.error_occurred.emit(str(e))


# ── Main UI ────────────────────────────────────────────────────────────────────

class IntroScreen(QtWidgets.QWidget):
    def __init__(self):
        super().__init__()

        # Start the local host server if available
        if hoster:
            target_path = os.path.join(_PROJECT_ROOT, "host", "styles.css")
            if os.path.exists(target_path):
                hoster.host(target_path)

        self.loaders: list = []
        self.init_ui()

        if self.server_input.text():
            self.fetch_covers()
        else:
            self.input_widget.show()

    def closeEvent(self, event):
        if hoster:
            hoster.stop_host()
        event.accept()

    def init_ui(self):
        self.setWindowTitle("Meteor")

        # ── Use C++ (or fallback) to compute a good window size ───────────────
        screen   = QtWidgets.QApplication.primaryScreen()
        geom     = screen.geometry()
        win_w, win_h = _ideal_window_size(geom.width(), geom.height(), 0.55)
        self.resize(win_w, win_h)

        self.main_layout = QtWidgets.QVBoxLayout(self)
        self.main_layout.setContentsMargins(20, 20, 20, 20)
        self.main_layout.setSpacing(15)

        # Server address input bar
        self.input_widget  = QtWidgets.QWidget()
        self.input_layout  = QtWidgets.QHBoxLayout(self.input_widget)
        self.input_layout.setContentsMargins(0, 0, 0, 0)

        self.server_input = QtWidgets.QLineEdit()
        self.server_input.setPlaceholderText("Server address (e.g. 127.0.0.1:8304)")

        self.connect_btn = QtWidgets.QPushButton("Fetch Covers")
        self.connect_btn.clicked.connect(self.fetch_covers)

        self.input_layout.addWidget(self.server_input)
        self.input_layout.addWidget(self.connect_btn)
        self.main_layout.addWidget(self.input_widget)
        self.input_widget.hide()

        # Cover collage
        self.scroll_area = QtWidgets.QScrollArea()
        self.scroll_area.setWidgetResizable(True)
        self.collage_container = QtWidgets.QWidget()
        self.collage_layout    = QtWidgets.QGridLayout(self.collage_container)
        self.collage_container.setLayout(self.collage_layout)
        self.scroll_area.setWidget(self.collage_container)
        self.main_layout.addWidget(self.scroll_area)

        # Auth buttons
        self.auth_layout = QtWidgets.QHBoxLayout()
        self.login_btn   = QtWidgets.QPushButton("Login")
        self.signup_btn  = QtWidgets.QPushButton("Sign Up")

        self.auth_layout.addStretch()
        self.auth_layout.addWidget(self.login_btn)
        self.auth_layout.addSpacing(20)
        self.auth_layout.addWidget(self.signup_btn)
        self.auth_layout.addStretch()
        self.main_layout.addLayout(self.auth_layout)

        self.load_config()

    def load_config(self):
        """
        Load server address from config.  Uses _read_config() which delegates
        to the C++ fast-path (meteor_cpp.read_config) when available.
        """
        config_path = os.path.join(_PROJECT_ROOT, "user_files", "config.json")
        config = _read_config(config_path)

        ip   = config.get("ip",   "")
        port = config.get("port", "")
        if ip:
            addr = f"{ip}:{port}" if port else ip
            self.server_input.setText(addr)

    def fetch_covers(self):
        server = self.server_input.text().strip()
        if not server:
            self.input_widget.show()
            return

        # ── Validate address via C++ (or Python fallback) ─────────────────────
        if not _validate_address(server):
            self.input_widget.show()
            QtWidgets.QMessageBox.warning(self, "Invalid address",
                                          f"'{server}' does not look like a valid server address.")
            return

        # Clear old covers
        for i in reversed(range(self.collage_layout.count())):
            widget = self.collage_layout.itemAt(i).widget()
            if widget:
                widget.setParent(None)
        self.loaders.clear()

        self.fetcher = CoverFetcher(server, self)
        self.fetcher.covers_found.connect(self.handle_covers_found)
        self.fetcher.error_occurred.connect(self.handle_fetch_error)
        self.fetcher.start()

    def handle_fetch_error(self, error_msg: str):
        self.input_widget.show()

    def handle_covers_found(self, urls: list):
        self.input_widget.hide()

        row, col  = 0, 0
        cols_per_row = 4

        for i, url in enumerate(urls):
            label = QtWidgets.QLabel("Loading…")
            label.setFixedSize(150, 225)
            label.setStyleSheet("border: 1px solid #555;")
            label.setAlignment(Qt.AlignmentFlag.AlignCenter)
            label.setScaledContents(True)

            self.collage_layout.addWidget(label, row, col)

            loader = ImageLoader(url, i, self)
            loader.image_loaded.connect(
                lambda pix, idx, lbl=label: self.update_image(pix, lbl))
            self.loaders.append(loader)
            loader.start()

            col += 1
            if col >= cols_per_row:
                col = 0
                row += 1

    def update_image(self, pixmap: QtGui.QPixmap, label: QtWidgets.QLabel):
        if not pixmap.isNull():
            label.setPixmap(pixmap.scaled(
                label.size(),
                Qt.AspectRatioMode.KeepAspectRatioByExpanding,
                Qt.TransformationMode.SmoothTransformation))


# ─────────────────────────────────────────────────────────────────────────────
if __name__ == "__main__":
    app = QtWidgets.QApplication(sys.argv)
    window = IntroScreen()
    window.show()
    sys.exit(app.exec())
